tmexp/create_bow.py

```diff
@@ -0,0 +1,218 @@
+        for blob_hash, feature_dict in blob_dict.items():
+            word_dict: Counter = Counter()
+            for feature in features:
+                if feature not in feature_dict:
+                    continue
+                word_dict.update(feature_dict[feature])
+            if word_dict:
+                num_blobs += 1
+                doc_freq.update(word_dict.keys())
+                word_counts[blob_hash] = word_dict
+        if not word_counts:
+            continue
+        file_bows: Dict[int, Dict[str, Any]] = {}
+        prev_ref = None
+        prev_blob = None
+        if topic_model == DIFF_MODEL:
+            prev_word_counts: Counter = Counter()
+        for ref in all_refs:
+            if ref in refs:
+                cur_blob = input_dict["files_info"][ref][file_path]["blob_hash"]
+                cur_word_counts = copy.deepcopy(word_counts[cur_blob])
+            else:
+                if prev_ref is None:
+                    continue
+                cur_blob = None
+                cur_word_counts = Counter()
+            if not (cur_blob is None and topic_model == HALL_MODEL):
+                prev_id = len(file_bows) - 1
+                if prev_blob == cur_blob:
+                    file_bows[prev_id]["refs"].append(ref)
+                else:
+                    file_bows[prev_id + 1] = {"refs": [ref]}
+                    if topic_model == HALL_MODEL:
+                        file_bows[prev_id + 1]["all"] = cur_word_counts
+                    elif topic_model == DIFF_MODEL:
+                        cur_word_counts.subtract(prev_word_counts)
+                        file_bows[prev_id + 1]["added"] = {
+                            word: count
+                            for word, count in cur_word_counts.items()
+                            if count > 0
+                        }
+                        file_bows[prev_id + 1]["removed"] = {
+                            word: -count
+                            for word, count in cur_word_counts.items()
+                            if count < 0
+                        }
+                        if cur_blob is None:
+                            prev_word_counts = Counter()
+                        else:
+                            prev_word_counts = word_counts[cur_blob]
+                    num_docwords += len(
+                        [count for count in cur_word_counts.values() if count != 0]
+                    )
+                    prev_blob = cur_blob
+            prev_ref = ref
+        bow[file_path] = file_bows
+
+    if min_word_frac > 0 or max_word_frac < 1:
+        logger.info("Used %d blobs to create documents." % num_blobs)
+        min_word_blob = int(min_word_frac * num_blobs)
+        max_word_blob = int(max_word_frac * num_blobs)
+        logger.info(
+            "Finding words that appear in less then %d blobs or more then %d blobs ..."
+            % (min_word_blob, max_word_blob)
+        )
+        blacklisted_words = set(
```

Comment:
Wen can remove the intermediate list here again:
```suggestion
        blacklisted_words = {
            word
            for word, count in doc_freq.items()
            if count < min_word_blob or count > max_word_blob
        }
```

Reply:
done